[c-c++] Added clangd support. Separated lsp server selection from backend.

cormacc opened the pull request:
Replaces: #12443
Closes: #12351 

Added support for `clangd` (to the existing `ccls` and `cquery`) as lsp backends to the c-c++ layer.
Updated the backend selection mechanism -- now aligned with the approach used in the python layer, suggested by @sdwolfz during an earlier PR review.

Specifically:
The existing mechanism is to set the `'c-c++-backend` layer variable to one of `nil`, `'rtags`, `'lsp-ccls` or `'lsp-cquery`.
The new/proposed mechanism is to set `'c-c++-backend` to one of `nil`, `'rtags` or `'lsp` and (optionally) set the `'c-c++-lsp-server` layer variable to one of `'clangd`, `'ccls` or `'cquery`. If unspecified, the server selection defaults to `'clangd`.

Re. default backend selection, I personally prefer/use `'ccls`, but opted to leave `'clangd` as the default, as it's the default supported by the upstream `lsp-mode` package and (arguably) easier to install. Use of `'cquery` is no longer recommended, as the project seems to have been abandoned, however support has been left in place for existing users with a functional installation.

Also opportunistically tidied up a few bits of configuration, including eliminating some company-related config that's now done automagically upstream by the `lsp-mode` package.

Thanks to @kvaneesh for some early feedback, and @auscompgeek for accomodating this PR. And @yyoncho and @maskray for all their work on `lsp-mode` and `ccls` respectively.

N.B. there's a clangd-specific find-other-file extension to the LSP which I haven't bound, as...
a) I haven't got it to work (clearly not reading the spec / calling it correctly).
b) I'm perfectly happy with projectile-find-other-file, and using ccls rather than clangd anyway
If anyone else is interested in taking a look while reviewing this PR, it's wrapped in interactive defuns: `c-c++/find-clangd-other-file` and `c-c++/peek-clangd-other-file`. If you do get it to work, it could be bound under `,go`, `,Go` OR replace the existing `,ga` keybinding.

yyoncho commented:
Code changes look good to me but I wonder whether we should drop `cquery` support at all since it is a dead project.

cormacc commented:
I think cquery support should probably be dropped as you say yyoncho, as it seems pretty much abandoned at this stage. Let's leave that for a separate PR though, as could muddy the waters here, and they're fairly separate pieces of work.

henriquenj commented:
Thanks for your work @cormacc! I was trying to hack my own version of clangd support but yours seems a lot more promising!

Unfortunately I couldn't get it to work. I applied your patch on top of f42b2db568446131ba5a52a6c33bd0dabd40f478 but every time I open a C++ file I get the following message `File mode specification error: (error Invalid command list)`. Investigating a little bit further, I enabled debug with `spacemacs/toggle-debug-on-error` and manually called `lsp` on a C++ buffer. This is the backtrace that was generated:

```
Debugger entered--Lisp error: (cl-assertion-failed ((seq-every-p (apply-partially (function stringp)) command) "Invalid command list"))
  cl--assertion-failed((seq-every-p (apply-partially (function stringp)) command) "Invalid command list" nil nil)
  lsp-resolve-final-function(lsp-clients--clangd-command)
  #f(compiled-function () #<bytecode 0x46108a75>)()
  lsp--matching-clients\?(#s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x46108a61>) :test\? #f(compiled-function () #<bytecode 0x46108a75>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x46108a81> :request-handlers #<hash-table equal 0/65 0x46108ba9> :response-handlers #<hash-table eql 0/65 0x46108bc9> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x46108fc1> :action-handlers #<hash-table equal 0/65 0x46109635> :major-modes (c-mode c++-mode objc-mode) :activation-fn nil :priority -1 :server-id clangd :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil))
  funcall(lsp--matching-clients\? #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x46108a61>) :test\? #f(compiled-function () #<bytecode 0x46108a75>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x46108a81> :request-handlers #<hash-table equal 0/65 0x46108ba9> :response-handlers #<hash-table eql 0/65 0x46108bc9> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x46108fc1> :action-handlers #<hash-table equal 0/65 0x46109635> :major-modes (c-mode c++-mode objc-mode) :activation-fn nil :priority -1 :server-id clangd :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil))
  (if (funcall pred it) (progn (setq result (cons it result))))
  (let ((it (car list))) (if (funcall pred it) (progn (setq result (cons it result)))))
  (while list (let ((it (car list))) (if (funcall pred it) (progn (setq result (cons it result))))) (setq it-index (1+ it-index)) (setq list (cdr list)))
  (let ((list list) (it-index 0)) (while list (let ((it (car list))) (if (funcall pred it) (progn (setq result (cons it result))))) (setq it-index (1+ it-index)) (setq list (cdr list))))
  (let (result) (let ((list list) (it-index 0)) (while list (let ((it (car list))) (if (funcall pred it) (progn (setq result (cons it result))))) (setq it-index (1+ it-index)) (setq list (cdr list)))) (nreverse result))
  -filter(lsp--matching-clients\? (#s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x4621d09d>) :test\? #f(compiled-function () #<bytecode 0x4621d0b5>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x4621d0c1> :request-handlers #<hash-table equal 0/65 0x4621ecb9> :response-handlers #<hash-table eql 0/65 0x4621efb1> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x4621f2a9> :action-handlers #<hash-table equal 0/65 0x4621f581> :major-modes (python-mode) :activation-fn nil :priority -1 :server-id pyls :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn #f(compiled-function (workspace) #<bytecode 0x4621ce9d>) :before-file-open-fn nil :initialized-fn #f(compiled-function (workspace) #<bytecode 0x4621cecd>) :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x460578b9>) :test\? #f(compiled-function () #<bytecode 0x460578d1>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 1/65 0x460578dd> :request-handlers #<hash-table equal 0/65 0x46057dc1> :response-handlers #<hash-table eql 0/65 0x460579a1> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x460579c1> :action-handlers #<hash-table equal 1/65 0x46057981> :major-modes (rust-mode rustic-mode) :activation-fn nil :priority -1 :server-id rls :multi-root nil :initialization-options ((omitInitBuild . t) (cmdRun . t)) :custom-capabilities nil :library-folders-fn #f(compiled-function (workspace) #<bytecode 0x46055d69>) :before-file-open-fn nil :initialized-fn #f(compiled-function (workspace) #<bytecode 0x46055d81>) :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x46059969>) :test\? #f(compiled-function () #<bytecode 0x46059981>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x4605998d> :request-handlers #<hash-table equal 0/65 0x460599ad> :response-handlers #<hash-table eql 0/65 0x460599cd> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x460671c1> :action-handlers #<hash-table equal 0/65 0x46066dc1> :major-modes (ruby-mode enh-ruby-mode) :activation-fn nil :priority -1 :server-id ruby-ls :multi-root t :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn #f(compiled-function (workspace) #<bytecode 0x46059865>) :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x46228dd9>) :test\? #f(compiled-function () #<bytecode 0x46228ded>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x46228df9> :request-handlers #<hash-table equal 0/65 0x462290d1> :response-handlers #<hash-table eql 0/65 0x4622a3b9> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x4622a6b1> :action-handlers #<hash-table equal 0/65 0x4622a9a9> :major-modes (html-mode sgml-mode mhtml-mode web-mode) :activation-fn nil :priority -4 :server-id html-ls :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn #f(compiled-function (w) #<bytecode 0x4607e5bd>) :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x4622fd01>) :test\? #f(compiled-function () #<bytecode 0x4622fd15>)) :ignore-regexps nil :ignore-messages ("readFile .*? requested by Vue but content not available") :notification-handlers #<hash-table equal 0/65 0x4622fd21> :request-handlers #<hash-table equal 0/65 0x4622fff9> :response-handlers #<hash-table eql 0/65 0x4623009d> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x462304c1> :action-handlers #<hash-table equal 0/65 0x462300bd> :major-modes nil :activation-fn #f(compiled-function (filename mode) #<bytecode 0x462241d5>) :priority -1 :server-id vls :multi-root t :initialization-options #f(compiled-function () #<bytecode 0x462241f9>) :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn #f(compiled-function (workspace) #<bytecode 0x46224219>) :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x45e06b81>) :test\? #f(compiled-function () #<bytecode 0x45e06b99>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 2/65 0x45e06ba5> :request-handlers #<hash-table equal 0/65 0x45e06e7d> :response-handlers #<hash-table eql 0/65 0x45e06f21> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x460871c1> :action-handlers #<hash-table equal 0/65 0x45e06f41> :major-modes (php-mode) :activation-fn nil :priority -1 :server-id iph :multi-root t :initialization-options #f(compiled-function () #<bytecode 0x4607e0ed>) :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x4608e975>) :test\? #f(compiled-function () #<bytecode 0x4608e989>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x4608e9b5> :request-handlers #<hash-table equal 0/65 0x4608e9d5> :response-handlers #<hash-table eql 0/65 0x460911c1> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x46090dc1> :action-handlers #<hash-table equal 1/65 0x4608e995> :major-modes (css-mode less-mode less-css-mode sass-mode scss-mode) :activation-fn nil :priority -1 :server-id css-ls :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x4624112d>) :test\? #f(compiled-function () #<bytecode 0x460919e9>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x4624114d> :request-handlers #<hash-table equal 0/65 0x46241425> :response-handlers #<hash-table eql 0/65 0x462414c9> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x462418c1> :action-handlers #<hash-table equal 0/65 0x46241d01> :major-modes nil :activation-fn #f(compiled-function (file-name mode) #<bytecode 0x4623d8ad>) :priority 0 :server-id xmlls :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn #f(compiled-function (workspace) #<bytecode 0x4623d8c5>) :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x462420e9>) :test\? #f(compiled-function () #<bytecode 0x460a22a9>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x460a22b5> :request-handlers #<hash-table equal 0/65 0x460a258d> :response-handlers #<hash-table eql 0/65 0x460a25ad> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x460a25cd> :action-handlers #<hash-table equal 0/65 0x460a4dc1> :major-modes (go-mode) :activation-fn nil :priority 0 :server-id gopls :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn #f(compiled-function (workspace) #<bytecode 0x46242011>) :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x460a5541>) :test\? #f(compiled-function () #<bytecode 0x460a5559>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x460a5565> :request-handlers #<hash-table equal 0/65 0x460a5585> :response-handlers #<hash-table eql 0/65 0x460a6ac1> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x460a6ec1> :action-handlers #<hash-table equal 0/65 0x460a55a5> :major-modes (go-mode) :activation-fn nil :priority -1 :server-id go-bingo :multi-root nil :initialization-options lsp-clients-go--make-init-options :custom-capabilities nil :library-folders-fn #f(compiled-function (workspace) #<bytecode 0x460a54cd>) :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x460a6aed>) :test\? #f(compiled-function () #<bytecode 0x460a7569>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x460a7575> :request-handlers #<hash-table equal 0/65 0x460a769d> :response-handlers #<hash-table eql 0/65 0x460a7ac1> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x460a76bd> :action-handlers #<hash-table equal 0/65 0x460a76dd> :major-modes (go-mode) :activation-fn nil :priority -2 :server-id go-ls :multi-root nil :initialization-options lsp-clients-go--make-init-options :custom-capabilities nil :library-folders-fn #f(compiled-function (workspace) #<bytecode 0x460a54d9>) :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x460a8681>) :test\? #f(compiled-function () #<bytecode 0x460a8699>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x460a86c5> :request-handlers #<hash-table equal 0/65 0x460b53c1> :response-handlers #<hash-table eql 0/65 0x460b57c1> :prefix-function nil :uri-handlers #<hash-table equal 1/65 0x460a86a5> :action-handlers #<hash-table equal 0/65 0x460b5b61> :major-modes (clojure-mode clojurec-mode clojurescript-mode) :activation-fn nil :priority 0 :server-id clojure-lsp :multi-root nil :initialization-options (:dependency-scheme "jar") :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x460b5f61>) :test\? #f(compiled-function () #<bytecode 0x460b5f75>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x460b5f81> :request-handlers #<hash-table equal 0/65 0x460b5fa1> :response-handlers #<hash-table eql 0/65 0x40b2de41> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x40b2e241> :action-handlers #<hash-table equal 0/65 0x460b5fc1> :major-modes (dart-mode) :activation-fn nil :priority -2 :server-id dart_language_server :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x40b2de6d>) :test\? #f(compiled-function () #<bytecode 0x40b2ea65>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x40b2ec31> :request-handlers #<hash-table equal 0/65 0x40b2ef29> :response-handlers #<hash-table eql 0/65 0x40b2f201> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x40b2f221> :action-handlers #<hash-table equal 0/65 0x40b2f641> :major-modes (dart-mode) :activation-fn nil :priority -1 :server-id dart_analysis_server :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x460c5089>) :test\? #f(compiled-function () #<bytecode 0x460c50a1>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x460c50ad> :request-handlers #<hash-table equal 0/65 0x460c5385> :response-handlers #<hash-table eql 0/65 0x460c53a5> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x460c53c5> :action-handlers #<hash-table equal 0/65 0x460c5cc1> :major-modes (elm-mode) :activation-fn nil :priority -1 :server-id elm-ls :multi-root nil :initialization-options lsp-clients-elm--make-init-options :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x460d4f19>) :test\? #f(compiled-function () #<bytecode 0x460d4f31>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 2/65 0x460d4f3d> :request-handlers #<hash-table equal 0/65 0x460d516d> :response-handlers #<hash-table eql 0/65 0x460d518d> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x460d51ad> :action-handlers #<hash-table equal 0/65 0x460d51cd> :major-modes (scala-mode) :activation-fn nil :priority -1 :server-id metals :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn #f(compiled-function (workspace) #<bytecode 0x460d4eb1>) :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x460e3ae9>) :test\? #f(compiled-function () #<bytecode 0x460e3b01>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 4/65 0x460e3b0d> :request-handlers #<hash-table equal 0/65 0x460e3e01> :response-handlers #<hash-table eql 0/65 0x460e40d9> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x460e417d> :action-handlers #<hash-table equal 0/65 0x460e45c1> :major-modes (fsharp-mode) :activation-fn nil :priority 0 :server-id fsac :multi-root nil :initialization-options lsp-fsharp--make-init-options :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x460f1541>) :test\? #f(compiled-function () #<bytecode 0x460f1559>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x460f1565> :request-handlers #<hash-table equal 0/65 0x460f1585> :response-handlers #<hash-table eql 0/65 0x460f19c1> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x460f1dc1> :action-handlers #<hash-table equal 0/65 0x460f15a5> :major-modes (erlang-mode) :activation-fn nil :priority -1 :server-id erlang-ls :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x460ff2bd>) :test\? #f(compiled-function () #<bytecode 0x460ff2d5>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 4/65 0x46102401> :request-handlers #<hash-table equal 0/65 0x461026d9> :response-handlers #<hash-table eql 0/65 0x4610277d> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x46102bc1> :action-handlers #<hash-table equal 0/65 0x4610279d> :major-modes (haxe-mode) :activation-fn nil :priority -1 :server-id haxe :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn #f(compiled-function (workspace) #<bytecode 0x460ff1d9>) :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x461032e9>) :test\? #f(compiled-function () #<bytecode 0x46103301>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x4610330d> :request-handlers #<hash-table equal 0/65 0x461033b1> :response-handlers #<hash-table eql 0/65 0x461033d1> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x4603bdc1> :action-handlers #<hash-table equal 0/65 0x46103ab9> :major-modes (sh-mode) :activation-fn nil :priority -1 :server-id bash-ls :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x46104171>) :test\? #f(compiled-function () #<bytecode 0x46104185>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x46104191> :request-handlers #<hash-table equal 0/65 0x461043c1> :response-handlers #<hash-table eql 0/65 0x46104761> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x46104781> :action-handlers #<hash-table equal 0/65 0x46104bc1> :major-modes (groovy-mode) :activation-fn nil :priority -1 :server-id groovy-ls :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x461053d5>) :test\? #f(compiled-function () #<bytecode 0x461053ed>)) :ignore-regexps nil :ignore-messages ("readFile .*? requested by TypeScript but content not available") :notification-handlers #<hash-table equal 0/65 0x40b2cee5> :request-handlers #<hash-table equal 0/65 0x40b2d1bd> :response-handlers #<hash-table eql 0/65 0x40b3b0b5> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x40b3b3ad> :action-handlers #<hash-table equal 0/65 0x40b3b5dd> :major-modes nil :activation-fn lsp-typescript-javascript-tsx-jsx-activate-p :priority -3 :server-id jsts-ls :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x40b3b661>) :test\? #f(compiled-function () #<bytecode 0x40b3b261>)) :ignore-regexps nil :ignore-messages ("readFile .*? requested by TypeScript but content not available") :notification-handlers #<hash-table equal 0/65 0x40b3b9e1> :request-handlers #<hash-table equal 0/65 0x40b3ba01> :response-handlers #<hash-table eql 0/65 0x40b3be41> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x40b3c241> :action-handlers #<hash-table equal 0/65 0x40b3ba21> :major-modes nil :activation-fn lsp-typescript-javascript-tsx-jsx-activate-p :priority -2 :server-id ts-ls :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x40b3ccd9>) :test\? #f(compiled-function () #<bytecode 0x40b3ccf1>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x40b3ccfd> :request-handlers #<hash-table equal 0/65 0x40b3ce25> :response-handlers #<hash-table eql 0/65 0x40b3ce45> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x46105fc1> :action-handlers #<hash-table equal 0/65 0x46106635> :major-modes nil :activation-fn lsp-clients-flow-activate-p :priority -1 :server-id flow-ls :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x461063e9>) :test\? #f(compiled-function () #<bytecode 0x40b3ce6d>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x46106991> :request-handlers #<hash-table equal 0/65 0x46106bc1> :response-handlers #<hash-table eql 0/65 0x46106f61> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x46106f81> :action-handlers #<hash-table equal 0/65 0x461073c1> :major-modes (php-mode) :activation-fn nil :priority -2 :server-id php-ls :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x461067ed>) :test\? #f(compiled-function () #<bytecode 0x46105fe1>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x46106fcd> :request-handlers #<hash-table equal 0/65 0x46107b61> :response-handlers #<hash-table eql 0/65 0x46107b81> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x46107fc1> :action-handlers #<hash-table equal 0/65 0x461083c1> :major-modes (reason-mode caml-mode tuareg-mode) :activation-fn nil :priority -1 :server-id ocaml-ls :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x46108a61>) :test\? #f(compiled-function () #<bytecode 0x46108a75>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x46108a81> :request-handlers #<hash-table equal 0/65 0x46108ba9> :response-handlers #<hash-table eql 0/65 0x46108bc9> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x46108fc1> :action-handlers #<hash-table equal 0/65 0x46109635> :major-modes (c-mode c++-mode objc-mode) :activation-fn nil :priority -1 :server-id clangd :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x46109a45>) :test\? #f(compiled-function () #<bytecode 0x46109a59>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x46109a65> :request-handlers #<hash-table equal 0/65 0x46109b8d> :response-handlers #<hash-table eql 0/65 0x46109fc1> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x46109bad> :action-handlers #<hash-table equal 0/65 0x46109bcd> :major-modes (elixir-mode) :activation-fn nil :priority -1 :server-id elixir-ls :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn #f(compiled-function (workspace) #<bytecode 0x46109999>) :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x4610ae65>) :test\? #f(compiled-function () #<bytecode 0x4610ae7d>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x4610ae89> :request-handlers #<hash-table equal 0/65 0x4610afb1> :response-handlers #<hash-table eql 0/65 0x4610afd1> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x4610bcc1> :action-handlers #<hash-table equal 0/65 0x4610c335> :major-modes (f90-mode fortran-mode) :activation-fn nil :priority -1 :server-id fortls :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x4610c691>) :test\? #f(compiled-function () #<bytecode 0x4610c6a9>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x4610c6b5> :request-handlers #<hash-table equal 0/65 0x4610f9a9> :response-handlers #<hash-table eql 0/65 0x4610fc81> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x4610fca1> :action-handlers #<hash-table equal 0/65 0x461100c1> :major-modes (kotlin-mode) :activation-fn nil :priority -1 :server-id kotlin-ls :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn #f(compiled-function (workspace) #<bytecode 0x4610acb5>) :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x461105a9>) :test\? #f(compiled-function () #<bytecode 0x461105bd>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 2/65 0x461105c9> :request-handlers #<hash-table equal 1/65 0x461108a1> :response-handlers #<hash-table eql 0/65 0x461108c1> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x46110cc1> :action-handlers #<hash-table equal 0/65 0x461110c1> :major-modes (hack-mode) :activation-fn nil :priority -1 :server-id hack :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x461114c5>) :test\? #f(compiled-function () #<bytecode 0x461114d9>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x46111501> :request-handlers #<hash-table equal 0/65 0x461117d9> :response-handlers #<hash-table eql 0/65 0x4611187d> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x46111cc1> :action-handlers #<hash-table equal 0/65 0x4611189d> :major-modes (dockerfile-mode) :activation-fn nil :priority -1 :server-id dockerfile-ls :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? t :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x46111ced>) :test\? #f(compiled-function () #<bytecode 0x461123ed>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x461123f9> :request-handlers #<hash-table equal 0/65 0x4611249d> :response-handlers #<hash-table eql 0/65 0x461128c1> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x461124bd> :action-handlers #<hash-table equal 0/65 0x461124dd> :major-modes nil :activation-fn #f(compiled-function (&rest args) #<bytecode 0x4610ad29>) :priority -1 :server-id angular-ls :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil) #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x46112ce5>) :test\? #f(compiled-function () #<bytecode 0x461128e1>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x461132dd> :request-handlers #<hash-table equal 0/65 0x46113489> :response-handlers #<hash-table eql 0/65 0x461134a9> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x461134c9> :action-handlers #<hash-table equal 0/65 0x46113cc1> :major-modes (plain-tex-mode latex-mode) :activation-fn nil :priority -1 :server-id digestif :multi-root nil :initialization-options nil :custom-capabilities nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil)))
  lsp--find-clients()
  lsp--try-project-root-workspaces(nil nil)
  lsp(nil)
```

My layer configuration

```
     ivy
     auto-completion
     better-defaults
     emacs-lisp
     git
     html
     markdown
     (ibuffer :variables ibuffer-group-buffers-by 'projects)
     themes-megapack
     yaml
     python
     multiple-cursors
     (lsp)
     (java :variables java-backend 'lsp)
     (c-c++ :variables c-c++-default-mode-for-headers 'c++-mode
            c-c++-backend 'lsp)
```

Do you have any idea what could be wrong on my setup? I'm on MacOS, but I have also tested on Ubuntu 16 with a clean installation.

cormacc commented:
Hi @henriquenj - I find I now get the same issue when I switch back to the clangd backend (I'm using ccls generally, which still works fine). At work at the moment, but will try and figure it out on my own time later. @yyoncho could this be a regression upstream in `lsp-mode`? All that's changed for me locally since I tested clangd last is updating lsp-mode package and rebasing my branch on develop tip (though rebasing hasn't entailed any overlapping changes in c-c++ or lsp layers as far as I'm aware).

yyoncho commented:
@cormacc I think that `lsp-clients--clangd-command` returns incorrect value, it should be list of strings. What does `M-: (lsp-clients--clangd-command) RET` returns?

cormacc commented:
Thanks @yyoncho  - that pointed me in the right direction. I had changed something -- added clangd specific handling of the layer c-c++-lsp-executable config option and was checking for symbol existence rather than value : S

Fixed / force pushed. @henriquenj  try again when you get a chance. Rebased on develop this morning, so you should be able to just checkout this branch altogether.

henriquenj commented on layers/+lang/c-c++/packages.el:
```diff
@@ -279,33 +280,30 @@
+    (unless (eq c-c++-lsp-server 'clangd)
+      (require c-c++-lsp-server) ; i.e. the 'ccls or 'cquery package
+      (remhash 'clangd lsp-clients))
+    (spacemacs/add-to-hooks 'lsp c-c++-mode-hooks)))
```
In the spirit of spacemacs of lazy loading everything, I suggest using `lsp-deferred` here instead of `lsp`, this will prevent clangd from starting if there's no C++ buffer on any window. @yyoncho What do you think?

yyoncho replied:
I think that it should be consistent across all lsp enabled layers, so IMO it should be pushed up in lsp layer and it should be configurable - thus it is out of scope for this PR.

henriquenj commented:
@cormacc Thanks for your new patch. I have tested it locally and it seems to be working now.

cormacc commented:
Unless otherwise noted via comment here, the regular force pushed updates are just me rebasing on develop rather than actual changes, as I'm working off this branch myself at the moment.

cormacc commented:
Rebased, then modified DAP config for compatibility with changes (DAP keybindings now done in dap layer config for all supported mode / per-mode call to spacemacs/dap-bind-keys-for-mode no longer required)

syl20bnr commented on layers/+lang/c-c++/layers.el:
```diff
@@ -10,5 +10,5 @@
-           (member c-c++-backend '(lsp-cquery lsp-ccls)))
-  (configuration-layer/declare-layer-dependencies '(lsp dap)))
+           (eq c-c++-backend 'lsp))
+  (configuration-layer/declare-layers '(lsp dap)))
```
Use `configuration-layer/declare-layer-dependencies` instead of `configuration-layer/declare-layers`.
If you encounter a bug using it, please let me know and I'll try to fix it.

cormacc replied:
Whoops. Fixed.

syl20bnr commented:
That's a great PR! I'll look into it later today.

cormacc commented:
My first stab at updated DAP configuration got it wrong (omitted adding cmodes to `spacemacs--dap-supported-modes` and requiring dap-lldb). Corrected now and amended/pushed again.
At least keybindings are appearing again :) I'm working on a cross-compiled embedded C codebase so have never actually used DAP (yet, anyway).

gessen commented on layers/+lang/c-c++/packages.el:
```diff
@@ -315,3 +314,8 @@
+(defun c-c++/pre-init-dap-mode ()
+  (dolist (mode c-c++-modes)
+    (add-to-list 'spacemacs--dap-supported-modes mode)
+    (when (eq c-c++-backend 'lsp) (spacemacs/add-to-hooks 'spacemacs/c-c++-setup-lsp-dap c-c++-mode-hooks))))
```
This function is written with double slash: spacemacs//c-c++-setup-lsp-dap

cormacc replied:
...ahem... :)
Thanks @gessen -- fixed/pushed.

syl20bnr commented:
Thank you ! :+1:

Cherry-picked into develop branch, *but not yet pushed to develop branch*.

The `c-c++` layer is very hard to follow, I'm currently refactoring it based on your PR to make it both simpler and more modern given the tools we have now in the ecosystem.

I'll need several days to finish this though and I have a nightly maintenance this week at my job so I expect to deliver the refactor on Friday.

In the meantime, can someone point me to a simple C++ project where I could test the various LSP backends ?

syl20bnr commented:
The reason why I don't push it yet is because the refactoring will also introduce breaking change in the layer configuration, so I prefer to avoid two big changes in a row. Sorry for the inconvenience but the C/C++ layer needs a big refresh.

yyoncho commented:
@syl20bnr 

> In the meantime, can someone point me to a simple C++ project where I could test the various LSP backends ?

Not exactly what you are looking for but I am using https://github.com/emacs-lsp/lsp-docker for testing the language servers. It has most of them preinstalled + test projects for some of the languages. For cpp it is "Hello world" project though. This is sufficient to test the spacemacs code but won't be sufficient if you want to evaluate which language server is better. lsp-docker also has a spacemacs configuration. It is still WIP but you might find it usefull.

cormacc commented:
Thanks @syl20bnr. The c-c++ layer is indeed in need of some love -- it was difficulties getting a consistent working setup across multiple dev environments that initially sent me down the lsp road.

Pretty much any cmake project should work to evaluate the lsp backends -- just need something that can generate a compile_commands.json. Though maybe better to stick with something that compiles for your host platform rather than getting stuck in the weeds with a cross-compiled project :)

Re. evaluating the various lsp backends, I'd focus on ccls and clangd. cquery appears to be abandoned -- the reasons I didn't strip support for it in this PR were:
1. It's a separate/atomic piece of work
2. Some users with an existing/working cquery setup may have been upset, potentially derailing this PR for a peripheral reason
3. As ccls is a fork of cquery and I originally added support for both in the same PR, there's a good deal of shared configuration in the layer. If/when removing cquery support it would make sense to flatten some of this -- those changes interleaved with the clangd ones may have made this PR more confusing to review.

syl20bnr commented:
Bear with me, I am still refactoring the layer. Flattening it out is pretty long process, it will indeed make it a lot easier to maintain afterwards. I'm maybe at 50% of the refactoring, maybe more.

syl20bnr commented:
https://github.com/syl20bnr/spacemacs/pull/12869 has been pushed to develop branch with this PR as the basis.

cormacc commented:
Great - thanks @syl20bnr !
